pkg/cloudprovider/provider/anexia/types/types.go

```diff
@@ -29,3 +42,10 @@ type RawConfig struct {
 	Memory     int                                 `json:"memory"`
 	DiskSize   int                                 `json:"diskSize"`
 }
+
+type ProviderStatus struct {
```

Comment:
Yeah I don't understand why you don't use this field since it's designed for this (even comment on the field says that, and status is correct place to keep information like this). Otherwise annotations sounds more correct for me to use than labels.

However, since I get the main argument here is to be sync with other providers I can understand that and change it. Although I don't see that any provider is using labels to keep the provider specific status. I see that labels are mostly used in context of metrics (location, region, type etc.) which is correct because labels are about organazing k8s objects.

I am not saying that is bad to put some ids to labels, my point is that we don't use ip, instance, and provisioning ids to in context of organizing objects but simply to identify resources on some other cloud provider. Different context. And even then I would probably keep both labels and status.

Just wanted to clarify how I see this.

Reply:
+1 for using `ProviderStatus`, IMO it's more appropriate than using annotations.

Reply:
Those are labels not annotations! @maciaszczykm Those are not statuses the thing that you use, those are names and ids, how they fit in the status object that I can't understand! They never change as simple as that!

Reply:
Btw I am @mfranczy not @maciaszczykm :)

Let me elaborate on that then, to make you understand my perspective I have to first explain how I understand spec and status that you get my point better.

For this specific example let me talk about newer API where spec-status split happened, not older where enabling /status endpoint was a breaking change (controllers were writing status to the main endpoint, not to /status endpoint).

So spec reflects a desired state of a resource where status summarize the current state of the object.

Instance ID (or any needed id) can reflect the current status of the object, if it's available it means VM has been created.
That represents a status of the Machine CR that corresponding VM has been created at some cloud provider,
either it's present - Machine has corresponding VM, or it's empty - Machine has not corresponding VM, then we must create one.

And in this case we only care about controller that it understands the current resource state.
For humans we could just add additional field with `Ready` or `NotReady` values.
But our priority is to make sure that controller knows what is the current state, and we need IDs to make it understandable as those are used to fetch information and reflect the status itself.
It's the same like with IP addresses field, there you have bunch of ips in the status field, and that represents the current network status too.

You say:

> They never change as simple as that!

and that's fine, they don't have to change as long they show valid status (but in that case change is from "" to "<id>" - and who or what says that it should change over time?).
For example, usually when you start VMs IPs don't change (but of course they could if you would reboot in case of dhcp - otherwise don't have to).

> Those are not statuses the thing that you use, those are names and ids, how they fit in the status object that

Of course names and IDs can represent status, in my opinion they can as i described above.

I think you're misleading `status` with `status.conditions` which is typical status field.

Things I said above brings me to another point, why I think it's bad to use labels as a reference to resources at some cloud providers.

By default `/status` subresource can be updated only by controller, probably you could change rbac and allow users to do it, but it's intended only for controllers to reflect the current state. This is crucial.

Labels can be easily changed by users, and as I said before those 
are only used to organize k8s objects and used as a reference to their relatives.

I think It's a bad idea to keep instance-id label as only source for the controller to fetch VM details at some cloud provider.

Imagine the following situation, 
1. we have two Machine CRs, they have different spec
1. two vms are being created based on that
1. we report ids of those vms to labels and this becomes primary source for the controller to fetch VM(!)

Suddenly someone makes mistake, or some other buggy controller updates labels and let's say change
id of one created VM to another one. This is what happens:

1. controller when checking the status of the vm will realize that expected state is different than the current one (for example first spec had different requested resources than the other one), so controller to fulfill requested expectation will start changing the VM.
1. you can lose information about original corresponding vm

Crucial information like that cannot be set in labels, because labels were not desgined for that.

So to summarize, status can contain any field that reflects the real state, and it can be done over IDs, names or typical status fields. By default it's only intended to be used by controllers which makes it safe.
In addition to that, a kind of field you use or name doesn't matter as long it shows the current status.

That's how I see this, if you think I am wrong then please point me where and I would like to read counterargument to this.
I am open to constructive criticism. Thanks.

Reply:
> Btw I am @mfranczy not @maciaszczykm :)

Yeah sorry about that I actually wanted to mention you but the suggestions showed the other Marcin and thought that was you.

> For this specific example let me talk about newer API where spec-status split happened, not older where enabling /status endpoint was a breaking change (controllers were writing status to the main endpoint, not to /status endpoint). 

I totally understand that, and this point is a bit irrelevant in regards of my opinion.

> So spec reflects a desired state of a resource where status summarize the current state of the object.

And that's the point here, a state is something that is constantly changing based on how the resource is being reconciled in its lifecycle: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podstatus-v1-core.

>Instance ID (or any needed id) can reflect the current status of the object, if it's available it means VM has been created. 

Implicitly yeah sure, but the thing that tells that this instance is created and running is the phase which is part of the status not of the spec.  

>For example, usually when you start VMs IPs don't change (but of course they could if you would reboot in case of dhcp - otherwise don't have to). 

We have a method for that, which is called Addresses and this is used exactly for this purpose ;) so this scenario was thought of before adding this cloud provider.

> I think you're misleading status with status.conditions which is typical status field.

Not really, conditions are part of the status, I am not sure why you raised this hint. 

>Imagine the following situation..

**This is actually not possible as `machine.spec.labels` are immutable.**

But even if they were: 
  
> Suddenly someone makes mistake, or some other buggy controller updates labels and let's say change
id of one created VM to another one.

This scenario is similar to, "if someone deletes the cluster, we should stop them", if someone goes rouge and changing things that he shouldn't be changing then I think this is not our problem, if a malfunction behaviour is observed. 

Looking at the cluster api, which this field was actually inspired by, as you can see over here https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha4/machine_types.go#L104-L184 no spec meta related stuff such as instance ids are listed over there. I mean look at your updateStatus function, it used only during the creation which is a very clear signal that this thing doesn't represent a state but it represent value that will never change :) 

Anyway, if you are insisting on adding this change(especially that this change doesn't break any running logic), I am fully okay with that. But this means, all cloud providers are kinda following a specific pattern which is using immutable labels which are part of the spec of a machine) and this specific cloud provider is following a total different pattern which kinda introduce some inconsistency between the cloud providers. **If you can bring up a case where the labels that are added on the machine spec is not enough then In totally agree with you, other than that, I think that those added functions and types can be replaced by three lines of code.**

Reply:
@mfranczy since the benefit of `ProviderStatus` is dubious, let' have refactorings and logic changes in dedicated PR (with preliminary discussion and maybe proposal of some kind). For the moment please stick to what other providers do.

Reply:
@moadqassem I have much more comments to this what you said but I will just respond to a few arguments which I have chosen, and rest I will reply in the proposal were we can continue the discussion.

> And that's the point here, a state is something that is constantly changing based on how the resource is being reconciled in its lifecycle: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podstatus-v1-core. 

State **doesn't have to constantly change** based on how the resource is being reconciled. Look at the node status and info section, especially on the `machine-id` [link](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#nodesysteminfo-v1-core). I think it reflects similar use case of reporting some IDs in the status. I will elaborate on this in the proposal.

> Also I can shutdown that instance it will still have an instance id but it won't represent the state anymore.

Then it's a bug which should be fixed in this PR, when instance is gone the controller should remove the value of instance id and replace it with an empty value. Next call controller would see this and start creating a new VM. Based on the state of instance-id the controller reacts. Still the general concept is kept.

> I mean look at your updateStatus function, it used only during the creation which is a very clear signal that this thing doesn't represent a state but it represent value that will never change.

As i said in the above response it's a bug. But still it's ok if any value doesn't change as long the value represents a real state that can be set upon a start. Again I refer to node status info [link](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#nodesysteminfo-v1-core).

and 

> This is actually not possible as machine.spec.labels are immutable.

What? You want me to use `machine.spec.labels`? Not `machine.labels`? On this I will elaborate in the proposal too. Even if I would use `machine.spec.labels` it doesn't change the thing that labels are bad for keeping provider specific information that are not in context of objects organization. And even later i guess this `machine.spec.labels` are being propagated somewhere to `<some resource>.labels` which becomes a mutable object.

> @mfranczy since the benefit of ProviderStatus is dubious, let' have refactorings and logic changes in dedicated PR (with preliminary discussion and maybe proposal of some kind). For the moment please stick to what other providers do.

@kron4eg ok, that makes sense, when I am done with anexia we can start discussion on potential refactoring in the machine controller based on opinions from the machine controller team.

@kron4eg @moadqassem @irozzo-1A thanks for review and discussion. I will use `labels` for now.

@moadqassem which I should use `machine.spec.labels` or `machine.labels`?

Reply:
btw @mfranczy I actually took a look at the client which is being used and I found something strange. In there, I saw a field called Tags https://github.com/anexia-it/go-anxcloud/blob/main/pkg/vsphere/search/name.go#L28, however that field is not being set anywhere if I am seeing this correctly. And why you didn't use that type(VM) instead of the info object? There is already a method attached which can get you the vm and by name https://github.com/anexia-it/go-anxcloud/blob/main/pkg/vsphere/search/name.go#L39. Any idea how those tags are being set?

Reply:
@moadqassem tags are not implemented yet.

> And why you didn't use that type(VM) instead of the info object? There is already a method attached which can get you the vm and by name

I don't want to start another discussion where we compare fetching vms by name (which returns list) or id and appropriate endpoints for this. Let's discuss this somwhere later, maybe in another PR or proposal.

@erdii at the beginning you followed hetzner provider (was it hetzner?), please use method `ByName` suggested by @moadqassem that we can merge the PR. Thanks.

Reply:
Just to be clear, we actually don't look up instance using names only, we actually lookup names and machine uids and both are unique. That's why I mentioned the tags here because that's what we do.